maap/config_reader.py

```diff
@@ -1,278 +1,305 @@
-        config_paths = list(map(self.__get_config_path, [os.path.dirname(config_file_path), os.curdir, os.path.expanduser("~"), os.environ.get("MAAP_CONF") or '.']))
-
-        for loc in config_paths:
-            try:
-                with open(loc) as source:
-                    self.__config.read_file(source)
-                    configfile_present = True
-                    break
-            except IOError:
-                pass
-        if not configfile_present:
-            raise IOError("No maap.cfg file found. Locations checked: " + '; '.join(config_paths))
-
+def _get_maap_api_host_url_scheme():
+    # Check if OS ENV override exists, used by dev testing
+    scheme = os.environ.get("MAAP_API_HOST_SCHEME", None)
+    if not scheme:
+        logger.debug("No url scheme defined in env var MAAP_API_HOST_SCHEME using https")
+        scheme = "https"
+    return scheme
+
+
+def _get_config_url(maap_host):
+    # This is added to remove the assumption of scheme specially for local dev testing
+    # also maintains backwards compatibility for user to use MAAP("api.maap-project.org")
+    base_url = urlparse(maap_host)
+    maap_api_config_endpoint = os.getenv("MAAP_API_CONFIG_ENDPOINT", "api/environment/config")
+    if not base_url.scheme == '':
+        if base_url.scheme not in ["https", "http"]:
+            raise ValueError(f"Provided scheme {base_url.scheme}. Only http or https schemes allowed for MAAP API Host")
+    # If the netloc is empty, that means that the url does not contain scheme:// and the url parser would put the
+    # hostname in the path section. See https://docs.python.org/3.11/library/urllib.parse.html#urllib.parse.urlparse
+    if base_url.netloc == '':
+        scheme = _get_maap_api_host_url_scheme()
+        url_components = namedtuple(typename='Components',
+                                    field_names=['scheme', 'netloc', 'path', 'query', 'fragments'])
+        config_url = urlunsplit(url_components(scheme=scheme, netloc=base_url.path, path=maap_api_config_endpoint,
+                                               query='', fragments=''))
+        # config_url = urlunparse(url_components(scheme=scheme, netloc=base_url.path, path=maap_api_config_endpoint,
+        #                                        query='', fragments=''))
+    else:
+        config_url = urljoin(maap_host, maap_api_config_endpoint)
+    return config_url
+
+
+def _get_api_root(config_url, config):
+    # Set maap api root to currently supplied maap host
+    api_root_url = urlparse(config.get("service").get("maap_api_root"))
+    config_url = urlparse(config_url)
+    url_components = namedtuple(typename='Components',
+                                field_names=['scheme', 'netloc', 'path', 'query', 'fragments'])
+    return urlunsplit(url_components(scheme=config_url.scheme, netloc=config_url.netloc, path=api_root_url.path,
+                                           query='', fragments=''))
+    
+def _get_client_config(maap_host):
```

Comment:
If you memoize this function, then there's no need to make `MaapConfig` a singleton (thus you can remove the added complexity) because it's this particular function that you're looking to invoke only once per unique `maap_host` value.

```suggestion
@functools.cache
def _get_client_config(maap_host: str):
```